### PR TITLE
[IMP] website_forum: populate with posts and comments

### PR DESCRIPTION
**Pull request by lm4649**
Populate a database with different quantities of forum posts and comments.The number of replies per question and comments per post (question or reply) are defined by weights.

Scripts originally created to test the performance of SQL queries accompanying the redesign of the forum website. Can be used for future website forum upgrades.

see odoo/odoo#123709
see odoo-dev/odoo#2645
see odoo/upgrade#4834

Task-3349374

Description of the issue/feature this PR addresses:

Current behavior before PR:

Desired behavior after PR is merged:

---
I confirm I have signed the CLA and read the PR guidelines at www.odoo.com/submit-pr

### CODE REVIEW

File: addons/website_forum/populate/forum_post.py [hunk 1]
```diff
@@ -0,0 +1,33 @@
...
+                    'forum_id': values['forum_id']})
+                    for _ in range(random.choices(*zip(*QA_WEIGHTS.items()))[0])]
+
+        return [
+            ('name', populate.constant('post_{counter}')),
+            ('forum_id', populate.randomize(forum_ids)),
+            ('child_ids', populate.compute(create_answers)),
+            ('last_activity_date', populate.randomize([
+                fields.Datetime.now(),
+                fields.Datetime.add(fields.Datetime.now(), hours=1)],
```
Comment (tde-banana-odoo) [inline addons/website_forum/populate/forum_post.py:31]: Is last activity in the future ?

File: addons/website_forum/populate/mail_message.py [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+                if nb_comments == 0:
+                    nb_comments = populate.random.choices(*zip(*CP_WEIGHTS.items()))[0]
+                    post_to_comment_id = res_ids[idx]
+                    idx += 1
+                nb_comments -= 1
+                yield {**values, 'res_id': post_to_comment_id}
+
+        return [
+            ('body', populate.constant('message_body_{counter}')),
+            ('message_type', populate.constant('comment')),
```
Comment (tde-banana-odoo) [inline addons/website_forum/populate/mail_message.py:39]: Would give subtype_id (to comment), to be close to "reality"

File: addons/website_forum/populate/mail_message.py [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+            for values in iterator:
+                if nb_comments == 0:
+                    nb_comments = populate.random.choices(*zip(*CP_WEIGHTS.items()))[0]
+                    post_to_comment_id = res_ids[idx]
+                    idx += 1
+                nb_comments -= 1
+                yield {**values, 'res_id': post_to_comment_id}
+
+        return [
+            ('body', populate.constant('message_body_{counter}')),
```
Comment (tde-banana-odoo) [inline addons/website_forum/populate/mail_message.py:38]: Could be great to add various authors, just to see for example the impact of fetching various avatars

**Review by tde-banana-odoo — COMMENTED** [review]
Zboing ! Some comments while passing by !

File: addons/website_forum/populate/mail_message.py (hunk 3 above)
Comment (tde-banana-odoo) [reply addons/website_forum/populate/mail_message.py:39]: Note that we could set an author already and fix the avatar afterwards :D seems weird to use create_uid. Author is there for that purpose.

File: addons/website_forum/populate/mail_message.py [hunk 4]
```diff
@@ -0,0 +1,44 @@
...
+
+    def _populate_factories(self):
+        mail_message_subtype = self.env.ref('mail.mt_comment')
+
+        def get_res_id(iterator, *args):
+            """Randomly assign messages to some populated posts and answers.
+            This makes sure these questions and answers get one to four comments.
+            """
+            question_ids = self.env.registry.populated_models['forum.post']
+            answers = self.env['forum.post'].search_read([('parent_id', 'in', question_ids)], ['id'])
```
Comment (tde-banana-odoo) [inline addons/website_forum/populate/mail_message.py:22]: Always a bit lost with generators... how many times will this be called ? Just to be sure we don't do 1 billion search_read ^^

File: addons/website_forum/populate/mail_message.py [hunk 5]
```diff
@@ -0,0 +1,44 @@
...
+
+CP_WEIGHTS = {1: 35, 2: 30, 3: 25, 4: 10}
+
+class Message(models.Model):
+    _inherit = 'mail.message'
+    _populate_dependencies = ['forum.post', 'res.partner']
+    _populate_sizes = {'small': 100, 'medium': 1000, 'large': 90000}
+
+    def _populate_factories(self):
+        mail_message_subtype = self.env.ref('mail.mt_comment')
```
Comment (tde-banana-odoo) [inline addons/website_forum/populate/mail_message.py:15]: 'comment_subtype' ?

**Review by tde-banana-odoo — COMMENTED** [review]
Zboing, some answers !

**Comment by tde-banana-odoo** [comment]
@robodoo delegate=@flch-odoo 

Once you are ready feel free to send :)

File: addons/website_forum/populate/mail_message.py (hunk 4 above)
Comment (lm4649) [reply addons/website_forum/populate/mail_message.py:22]: `self.env['forum.post'].search_read([('parent_id', 'in', question_ids)], ['id']) `is called one time.

File: addons/website_forum/populate/forum_post.py [hunk 6]
```diff
@@ -0,0 +1,29 @@
...
+    _populate_dependencies = ['forum.forum']
+
+    def _populate_factories(self):
+        forum_ids = self.env.registry.populated_models['forum.forum']
+
+        def create_answers(values=None, random=None, **kwargs):
+            return [fields.Command.create({
+                    'name': f"reply to {values['name']}",
+                    'forum_id': values['forum_id']})
+                    for _ in range(random.choices(*zip(*QA_WEIGHTS.items()))[0])]
```
Comment (flch-odoo) [inline addons/website_forum/populate/forum_post.py:23]: Sorry if we're circling back to an already closed topic, but can't we do something to have different create_date for posts, answers, and comments?

**Review by flch-odoo — COMMENTED** [review]
Hi @lm4649 ,
I'm here to discuss these points at your best convenience so we can move on :)

File: addons/website_forum/populate/forum_post.py (hunk 6 above)
Comment (lm4649) [reply addons/website_forum/populate/forum_post.py:29]: by setting _log_access to False, you can define the records create date. Here is a try. You might want to look at the datetime ranges. point of attention Post._compute_relevancy may raise an error.

File: addons/website_forum/populate/mail_message.py (hunk 3 above)
Comment (lm4649) [reply addons/website_forum/populate/mail_message.py:52]: Not sure. I now get the author ids from the populated users. Took this opportunity to set the create uid as well.

File: addons/website_forum/populate/mail_message.py [hunk 7]
```diff
@@ -0,0 +1,81 @@
...
+            ('subtype_id', populate.constant(comment_subtype.id)),
+            ('write_date', populate.compute(get_write_date)),
+            ('write_uid', populate.compute(get_write_uid)),
+        ]
+
+    def _populate(self, size):
+        records = super()._populate(size)
+
+        _logger.info('forum.post: update last_activity_date for posts with comments')
+        comments = self._read_group([('id', 'in', records.ids)], ['res_id'], ['create_date:max'])
```
Comment (flch-odoo) [inline addons/website_forum/populate/mail_message.py:74]: Doesn't this leave out the comments on answers?

How about making a mapping of comment res_id to create date, then iterate over each **question** `q` and set `q.last_activity_date` to  the max of `q.answer_ids.mapped(lambda a: comments_to_date.get(a.id, q.last_activity_date))`?

**Review by flch-odoo — COMMENTED** [review]
Keep it up, we'll get there :) !

**Review by flch-odoo — COMMENTED** [review]
Hello @lm4649,
Thanks for the updates!
One change has a longer diff but no change in logic, probably ^^.
I'm finishing populating a LARGE one at home, It all goes well, we'll finish this one very soon.

File: addons/website_forum/populate/forum_post.py [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+            """Create random number of answers
+            We set the `last_activity_date` to convert it to `create_date` in `_populate`
+            as the ORM prevents setting these.
+            """
+            return [
+                fields.Command.create({
+                    'name': f"reply to {values['name']}",
+                    'forum_id': values['forum_id'],
+                    'content': 'Answer content',
+                    'last_activity_date': fields.Datetime.add(values['last_activity_date'], hours=populate.random.choice(hours)),
```
Comment (flch-odoo) [inline addons/website_forum/populate/forum_post.py:32]:
```suggestion
                    'last_activity_date': fields.Datetime.add(values['last_activity_date'], hours=random.choice(hours)),
```

File: addons/website_forum/populate/forum_post.py [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+
+class Post(models.Model):
+    _inherit = 'forum.post'
+    # Include an additional average of 2 post answers for each given size
+    # e.g.: 100 posts as populated_model_records = ~300 actual forum.posts records
+    _populate_sizes = {'small': 100, 'medium': 1000, 'large': 90000}
+    _populate_dependencies = ['forum.forum', 'res.users']
+
+    def _populate_factories(self):
+        forum_ids = self.env.registry.populated_models['forum.forum']
```
Comment (flch-odoo) [inline addons/website_forum/populate/forum_post.py:19]: We can here define a seeed for random, so we always populate the same way. 
See https://github.com/odoo/odoo/pull/139150/files#diff-02e881d3b2b12393dd0ad71209a4942c99b68011f64903c5ecb1c33046fa8f16R16

File: addons/website_forum/populate/mail_message.py [hunk 10]
```diff
@@ -0,0 +1,105 @@
...
+
+    def _populate_factories(self):
+        comment_subtype = self.env.ref('mail.mt_comment')
+
+        def get_author_id(iterator, *args):
+            user_ids = self.env.registry.populated_models['res.users']
+            author_ids = self.env['res.users'].search_read([('id', 'in', user_ids)], ['partner_id'])
+
+            for values in iterator:
+                author_id = populate.random.choice(author_ids)
```
Comment (flch-odoo) [inline addons/website_forum/populate/mail_message.py:24]: Same here?

File: addons/website_forum/populate/forum_post.py [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+        records = super()._populate(size)
+        user_ids = self.env.registry.populated_models['res.users']
+
+        # Overwrite auto-fields: use last_activity_date to update create date
+        _logger.info('forum.post: update create date and uid')
+        question_ids = tuple(records.ids)
+        query = """
+            UPDATE forum_post
+               SET create_date = last_activity_date,
+                   create_uid = floor(random() * (%(max_value)s - %(min_value)s + 1) + %(min_value)s)
```
Comment (flch-odoo) [inline addons/website_forum/populate/forum_post.py:62]: maybe there is a way to seed this one?

**Review by flch-odoo — COMMENTED** [review]
Hey @lm4649 ,
Final thing, then we merge, today!

**Comment by flch-odoo** [comment]
Thank you for your perseverance, @lm4649 !

@robodoo r+

**Comment by seb-odoo** [comment]
Hello, this is breaking the conditions on the populate of mail/discuss, because you completely override the setup of message, in particular `_populate_dependencies` and `_populate_sizes`.

Moreover, extension of `_populate_factories` really don't work well together, I feel like any decently complex case should just use `_populate` overrides instead.

I guess I'll do the fix in my PR.

**Comment by flch-odoo** [comment]
Hello @seb-odoo ,

I'm sorry we overlooked this inheritance problem for other apps installed with forum.

Ideally, we'd like to populate messages with these parameters after other populate ran, but as I think you said, there is no way to do this conveniently, right?
With slides_forum we used how we can change fields, and sizes, but it is not feasible for more complex situations.  

That said, we cannot conveniently use _populate to create our records either, so perhaps the tooling just isn't there to populate for all apps together...

**Comment by seb-odoo** [comment]
Hello, you can have a look here: https://github.com/odoo/odoo/pull/139269

I believe I have done the conversion while keeping full functionality.

For me `_populate_factories` is good for very simple cases, but it gets in the way for more complex setup.
The message populate for `form.post` is much simpler with just writing it manually in `_populate` rather than trying to use `_populate_factories` stuff.

Let me know if it still provides the expected result for you.

**Pull request closed without merging**